client/src/game/reducers/cardDeductionReducer.ts

```diff
@@ -0,0 +1,144 @@
+  hands.filter((hand) => hand !== hands[metadata.ourPlayerIndex])
+    .forEach((hand) => { calculateHandPossibilities(hand, hands, newDeck, cardCountMap); });
+  return newDeck;
+};
+
+const calculateHandPossibilities = (
+  handToCalculate: readonly number[],
+  hands: ReadonlyArray<readonly number[]>,
+  deck: CardState[],
+  cardCountMap: readonly number[][],
+) => {
+  const cardCountMapForHand = Array.from(
+    cardCountMap,
+    (arr) => Array.from(arr),
+  );
+  const unknownCards: number[] = [];
+  hands.forEach((hand) => hand.forEach((o) => {
+    const deckCard = deck[o];
+    // If card is not fully known then we need to run through the permutations.
+    // Otherwise, it's a card in another hand and we know we don't have that copy
+    // (and can reduce the count accordingly from the perspective of this hand).
+    if (deckCard.suitIndex === null || deckCard.rank === null || hand === handToCalculate) {
+      unknownCards.push(o);
+    } else {
+      cardCountMapForHand[deckCard.suitIndex][deckCard.rank] -= 1;
+    }
+  }));
+  handToCalculate.forEach((o) => {
+    const card = deck[o];
+    const possibilities = unknownCards
+      .filter((ro) => ro !== o)
+      .map((ro) => deck[ro].possibleCardsFromDeduction);
+    deck[o] = {
+      ...card,
+      possibleCardsFromDeduction: card.possibleCardsFromDeduction.filter(
+        (possibility) => possibilityValid(possibility, possibilities, 0, cardCountMapForHand),
+      ),
+    };
+  });
+};
+
+const possibilityValid = (
+  [suit, rank]: readonly [number, number],
+  possibilities: ReadonlyArray<ReadonlyArray<readonly [number, number]>>,
+  index: number,
+  cardCountMap: readonly number[][],
+) => {
+  if (possibilities.length === index) {
+    return cardCountMap[suit][rank] > 0;
+  }
+  // Avoiding duplicating the map for performance, so trying to undo the mutation as we exit
+  cardCountMap[suit][rank] -= 1;
+  if (
+    cardCountMap[suit][rank] >= 0
+    && possibilities[index].some(
+      (possibility) => possibilityValid(possibility, possibilities, index + 1, cardCountMap),
+    )
+  ) {
+    cardCountMap[suit][rank] += 1;
+    return true;
+  }
+  cardCountMap[suit][rank] += 1;
+  return false;
+};
+
+let cachedVariantName: string | null = null;
+let cachedCardCountMap: number[][] = [];
+
+const getCardCountMap = (
+  variant: Variant,
+) => {
+  if (variant.name === cachedVariantName) {
+    return Array.from(cachedCardCountMap, (arr) => Array.from(arr));
+  }
+
+  const possibleSuits: number[] = variant.suits.slice().map((_, i) => i);
+  const possibleRanks: number[] = variant.ranks.slice();
+  const possibleCardMap: number[][] = [];
+
+  possibleSuits.forEach((suitIndex) => {
+    possibleCardMap[suitIndex] = [];
+    const suit = variant.suits[suitIndex];
+    possibleRanks.forEach((rank) => {
+      possibleCardMap[suitIndex][rank] = deckRules.numCopiesOfCard(
+        suit,
+        rank,
+        variant,
+      );
+    });
+  });
```

Comment:
Can this be simplified a bit?
```suggestion
  const possibleCardMap: number[][] = [];

  variant.suits.forEach((suit, suitIndex) => {
    possibleCardMap[suitIndex] = [];
    variant.ranks.forEach((rank) => {
      possibleCardMap[suitIndex][rank] = deckRules.numCopiesOfCard(
        suit,
        rank,
        variant,
      );
    });
  });
```

Reply:
There's a very specific reason why it's the way it is now - if you don't forEach it will skip empty fields, and this has to be aligned with 1-5 and 7 (skipping 0 and 6). Try it out and you'll see :)

Reply:
Oh wow good point

Reply:
Thanks rob, I'll switch it back.